src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,201 @@
+        internal Dictionary<string, double> DoublePropertyValues = new Dictionary<string, double>();
+        internal Dictionary<string, Guid> GuidPropertyValues = new Dictionary<string, Guid>();
+        internal Dictionary<string, DictionaryTypeEnum> PropertiesKeys = new Dictionary<string, DictionaryTypeEnum>();
+
+        /// <summary>
+        /// Name of the telemetry event, should be unique between events.
+        /// </summary>
+        public abstract string Name { get; set; }
+
+        /// <summary>
+        /// Properties which describe the event.
+        /// </summary>
+        public virtual IDictionary<string, object> Properties
+        {
+            get
+            {
+                var currentBag = new Dictionary<string, object>();
+                foreach (var kvp in PropertiesKeys)
+                {
+                    switch (kvp.Value)
+                    {
+                        case DictionaryTypeEnum.String:
+                            currentBag.Add(kvp.Key, PropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Long:
+                            currentBag.Add(kvp.Key, LongPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Double:
+                            currentBag.Add(kvp.Key, DoublePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Bool:
+                            currentBag.Add(kvp.Key, BoolPropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.DateTime:
+                            currentBag.Add(kvp.Key, DateTimePropertyValues[kvp.Key]);
+                            break;
+                        case DictionaryTypeEnum.Guid:
+                            currentBag.Add(kvp.Key, GuidPropertyValues[kvp.Key]);
+                            break;
+                    }
+                }
+                return currentBag;
+            }
+        }
+
+        /// <summary>
+        /// Gets the <see cref="DataClassification"/> mapping to keys in <see cref="Properties"/>.
+        /// </summary>
+        public virtual IDictionary<string, DataClassification> PropertyDataClassification
+        {
+            get
+            {
+                return DataClassificationMapping;
+            }
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        public virtual void SetProperty(
+            string key,
+            string value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            PropertyValues[key] = value;
```

Comment:
>PropertyValues[key] = value; [](http://example.com/codeflow?start=12&length=28)

ArgumentNullException will be thrown if key is null.

Reply:
Added null checks